tokio/src/sync/watch.rs

```diff
@@ -412,18 +417,30 @@ mod state {
         }
 
         /// Load the current value of the state.
+        ///
+        /// Only used by the receiver and for debugging purposes.
+        ///
+        /// The receiver side (read-only) uses `Acquire` ordering for a proper handover
+        /// of the shared value with the sender side (single writer). The state is always
+        /// updated after modifying and before releasing the (exclusive) lock on the
+        /// shared value.
         pub(super) fn load(&self) -> StateSnapshot {
-            StateSnapshot(self.0.load(SeqCst))
+            StateSnapshot(self.0.load(Ordering::Acquire))
         }
 
         /// Increment the version counter.
-        pub(super) fn increment_version(&self) {
-            self.0.fetch_add(STEP_SIZE, SeqCst);
+        pub(super) fn increment_version_after_updating_shared_value_while_locked(&self) {
+            // Use `Release` ordering to ensure that the shared value
+            // has been written before updating the version. The shared
+            // value is still protected by an exclusive lock during this
+            // method.
+            self.0.fetch_add(STEP_SIZE, Ordering::Release);
         }
 
         /// Set the closed bit in the state.
         pub(super) fn set_closed(&self) {
-            self.0.fetch_or(CLOSED_BIT, SeqCst);
+            // Relaxed ordering is sufficient here.
+            self.0.fetch_or(CLOSED_BIT, Ordering::Relaxed);
```

Comment:
Let's keep release here. Even if it doesn't matter for the watch channel itself, it still establishes happens-before relationships for the user of the watch channel. With release, we ensure that anything that happens before dropping the sender is visible to callers of `has_changed` or `changed` that get a closed error.
```suggestion
            self.0.fetch_or(CLOSED_BIT, Ordering::Release);
```